### PR TITLE
Support Dynamic Collection Initializers

### PR DESCRIPTION
**Pull request by 333fred**
As part of support DynamicInvocations in collection initializers, I
fixed a couple of bugs in C#:
1) Nested collection initializers with a missing Set method would
generate an IInvalidOperation when there was nothing invalid in the
tree.
2) DynamicMemberReferences weren't being generated for dynamic
collection Add calls, so you couldn't tell what method was getting
called by the invocation.

Addressed a small followup from #27126.

Tag @dotnet/roslyn-compiler @AlekseyTs @dotnet/analyzer-ioperation for review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Nested collection initializers use the Get method, as they're calling Add on the collection, rather than setting the field. If the setter was missing, it would generate an InvalidOperation.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_Latebound.vb
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_Latebound.vb]: Followup from https://github.com/dotnet/roslyn/pull/27126#issuecomment-392873966. IOperation test hook passed locally on all VB compiler tests.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >// In collection initializers, the property is not actually set. Instead, it is retrieved for a series of Add methods, so we need [](start = 24, length = 129)

I think this should also apply to the case when a property is initialized with a nested initializer. Is this scenario handled properly?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >implicitReceiver.Type [](start = 111, length = 21)

I think type should be null here, it is definitely not the same as the type of the receiver

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >implicitReceiver.ConstantValue [](start = 111, length = 30)

Similarly, should be default, I think.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >boundAssignmentOperator.Right.Kind == BoundKind.CollectionInitializerExpression [](start = 90, length = 79)

What if the right is regular object initializer?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >inCollectionInitializer [](start = 126, length = 23)

The name feels somewhat misleading, the target isn't really inside a collection initializer, it is being initialized with a collection initializer.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: >Type: C1 [](start = 132, length = 8)

This doesn't look right.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: For consistency with other DynamicMemberReferences, I'll make this the DynamicType

---
In reply to: [192265008](https://github.com/dotnet/roslyn/pull/27308#discussion_r192265008) [](ancestors = 192265008)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >BoundObjectInitializerMember [](start = 71, length = 28)

Do we ever expect to run into this node outside of the context of an object initializer? If not, then we should probably make new parameter required and simply remove the case handling this node in the switch at the top of this file. We should also make sure the right thing happens when someone requests IOperation node for the corresponding syntax directly, rather than getting to it through the parent.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >CreateBoundObjectInitializerMemberOperation((BoundObjectInitializerMember)boundAssignmentOperator.Left, inCollectionInitializer: true) [](start = 22, length = 134)

Is this call bypassing the cache? Is it possible to end up with two distinct IOperation nodes for the target?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: > For consistency with other DynamicMemberReferences, I'll make this the DynamicType

That would make sense only if there is a separate dynamic invocation for this node, i.e. two dynamic invocations in IL rather than one. Please check IL, and if there is only one dynamic invocation, the type should be null because this is not a value from compiler's point of view

---
In reply to: [192269007](https://github.com/dotnet/roslyn/pull/27308#discussion_r192269007) [](ancestors = 192269007,192265008)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Yes, we'll see it for top-level initializers.

---
In reply to: [192269046](https://github.com/dotnet/roslyn/pull/27308#discussion_r192269046) [](ancestors = 192269046)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: It is bypassing the cache, but because we fetch IOperations by getting the enclosing statement and walking down, we shouldn't end up with 2 nodes here.

---
In reply to: [192269300](https://github.com/dotnet/roslyn/pull/27308#discussion_r192269300) [](ancestors = 192269300)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [393718591](https://github.com/dotnet/roslyn/pull/27308#issuecomment-393718591) [](ancestors = 393718591)

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 2)

**Comment by 333fred** [comment]
@dotnet-bot retest windows_release_unit32_prtest please

**Comment by 333fred** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler I need a second review please.

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review, need second review

**Comment by AlekseyTs** [comment]
@mavasani, @dotnet/analyzer-ioperation Please review

**Review by mavasani — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (agocke) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: This is a very long line. Consider splitting it.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs
Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: Why does this implement `IEnumerable<int>`? I think I'm missing something about this test case.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: Because I'm using a nested collection initializer to initialize it.

---
In reply to: [192830340](https://github.com/dotnet/roslyn/pull/27308#discussion_r192830340) [](ancestors = 192830340)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Will do in a follow up. Doesn't look long on a 4K monitor :).

---
In reply to: [192826248](https://github.com/dotnet/roslyn/pull/27308#discussion_r192826248) [](ancestors = 192826248)

**Comment by 333fred** [comment]
@agocke I think I addressed your comments.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs
Comment (agocke) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: Got it. Thanks.

**Review by agocke — APPROVED** [review]
LGTM

**Comment by 333fred** [comment]
@dotnet-bot retest windows_debug_spanish_unit32_prtest please

**Comment by 333fred** [comment]
Test failures were from the test systems disconnecting, fyi @dotnet/roslyn-infrastructure. Merging as this was just a rebase, and previous pr builds have passed.

**Pull request merged**
